### PR TITLE
feat(lsp): support `on_notify` client notification hooks

### LINKED ISSUE #15275: LSP client notification hooks
### Feature description

It would be very convenient to have, similar to `on_init` and `on_exit`, an `on_notify` hook that would be a table mapping from client-to-server notification methods to functions that would be called _immediately after_ the notification is sent. We could also have a "default" key in this table for a function that is called regardless of the particular method.

VSCode has the (seemingly undocumented) `middleware` option which does this -- see, for example, [how it is used in the Lean 4 VSCode extension to trigger on `textDocument/didChange`](https://github.com/leanprover-community/vscode-lean4/blob/d5f1926be34e73af23d551be779a5b98e721e0ba/vscode-lean4/src/leanclient.ts#L122), which is what we'll be trying to replicate at Julian/lean.nvim#84.

### PR DESCRIPTION
**Pull request by rish987**
Closes #15275.

### CODE REVIEW

**Comment by mjlbach** [comment]
A couple thoughts:

* on_notify is a bit ambiguous as you can have notifications from the server to the client as well (maybe on_client_notification?
* I'm not sure how I feel about the 0 index, I think I'd rather have a "default" key
* I don't particularly see the utility of this compared to just having a handler driven event system, but I read the referenced lean PR and it seems like something you all want

**Comment by rish987** [comment]
- done
- done
- What we need is a way to hook a function _after_ we've guaranteed that a certain notification has been sent in particular this is useful if that function makes a request to the server. I couldn't find an existing way to do this in general. We could piece something together for notifications that we send ourselves, but for "built-in" notifications (e.g. `textDocument/didChange`) that neovim handles, I don't see an easy way. Though note that I'm still speaking for myself at this point 🤣 Would you mind describing a bit more what you mean by handler driven event system?

**Comment by mjlbach** [comment]
I mean, you *need* to have events triggered on notifications from client -> server rather than on responses to those notifications (which are sent to the appropriate handlers)

**Comment by mfussenegger** [comment]
Wouldn't it be possible to use `on_lines` or `on_bytes` of `nvim_buf_attach` for this?

Otherwise you could already accomplish something like this by monkey patching the `notify` method of the `client` within `on_init`. (Although I wouldn't really recommend it)

```lua
local function on_init(client)
  local orig_notify = client.notify
  client.notify = function(method, params)
    orig_notify(method, params)
  end
end
```


We also had discussions in the past about a generic listener system for any lsp communication, and I think I'd rather prefer that generic listener system instead of something that is specific to notifications.

**Comment by rish987** [comment]
> I mean, you _need_ to have events triggered on notifications from client -> server rather than on responses to those notifications (which are sent to the appropriate handlers)?

This would indeed be ideal, but CMIIW it doesn't look like servers need to give responses to `textDocument/didChange` or notifications in general? And in fact ["a notification message must not send a response back"](https://microsoft.github.io/language-server-protocol/specifications/specification-3-17/#notificationMessage).

> For your use case, you could just set up your own on_lines callback, although there is the caveat that we are internally debouncing text changes.

Yes we could, but this just doesn't seem "correct" to me because it's not explicit when this is happening relative to nvim sending `textDocument/didChange`.

**Comment by rish987** [comment]
> We also had discussions in the past about a generic listener system for any lsp communication, and I think I'd rather prefer that generic listener system instead of something that is specific to notifications.

This indeed sounds like the way to go, I could extend this to other communications as well, though this would likely involve additional indexing in the table for whether something is coming from the client or the server and whether it is a request or notification.

**Comment by mjlbach** [comment]
Yes, so in this case you need something that subscribes to the notification, not something that can be driven by a handler (which was what I wanted to clarify) as you're using textDocument/didChange

Do we need to distinguish between requests/notifications for the API? I feel as though we could just have a generic post request/post notification callback, and then plugin authors could just match on the name of the request/notification they wanted. 

Likewise for notifications from the server, so just expose 2 hooks (incoming/outgoing)

**Comment by rish987** [comment]
^ done, let me know your thoughts on the api/documentation/code.

File: runtime/lua/vim/lsp/rpc.lua [hunk 1]
```diff
@@ -312,14 +312,27 @@ end
...
     cmd = { cmd, 's' };
     cmd_args = { cmd_args, 't' };
+    _on_server = { _on_server, 'f', true };
+    _on_client = { _on_client, 'f', true };
     dispatchers = { dispatchers, 't', true };
   }
 
+  local on_server = function(type, ...)
+    if _on_server then
+      pcall(_on_server, type, ...)
```
Comment (rish987) [inline runtime/lua/vim/lsp/rpc.lua:327]: Let me know your thoughts on the best way to report an error here.

**Comment by rish987** [comment]
Update: see Julian/lean.nvim#113 for how we'll be using this, specifically [here](https://github.com/Julian/lean.nvim/blob/b75de4120151222ac63c7ddf97438ded8e673119/lua/lean/init.lua#L38) and [here](https://github.com/Julian/lean.nvim/blob/b75de4120151222ac63c7ddf97438ded8e673119/lua/lean/_util.lua#L59) for how we'll be using this to exactly synchronize our state (in this case, the locations of a set of pins) with the file state on the server.

**Comment by rish987** [comment]
Julian/lean.nvim#113 is pretty much ready to go at this point! The main point here is to be able to update our infoview on `textDocument/didChange`, since doing so `on_lines` [would trigger `textDocument/didChange`](https://github.com/neovim/neovim/blob/222cd43222c68ba5376b681d2a5fd7b90da138b0/runtime/lua/vim/lsp.lua#L893), which we don't want when debouncing. @mjlbach @mfussenegger let me know what you think of the API here, I tried to make it "smart" and pass in the type of message (notification, request, response) and corresponding parameters based on what `rpc.lua` infers.

**Comment by rish987** [comment]
UPDATE: after a bit of thinking out loud at Julian/lean.nvim#115, I think we could just handle this at its root and make it configurable at `client.request` whether to postpone a request to the actual `textDocument/didChange` notification when there are pending debounced changes (rather than immediately flushing and making the request). Let me know what you think, will try this out and make a PR soon.

**Comment by rish987** [comment]
^ however, this still leaves it unclear in the case of a change when NOT debouncing, whether a request is triggered before or after `textDocument/didChange` when using `set_lines`. For that reason something like this seems necessary _in addition to_ the ability to postpone when debouncing. I could also see something as general as this being useful for example when developing a language server.

**Comment by rish987** [comment]
So, I've realized that we've gotten around the issue at Julian/lean.nvim#113 on our side with Julian/lean.nvim#114, at least when not debouncing, which allows us to use `on_lines`, so I'll close this for now. But if someone else comes up with the need for this feel free to revive it.

**Pull request closed without merging**